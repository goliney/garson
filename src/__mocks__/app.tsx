import React from 'react';
import { render } from 'ink-testing-library';
import { Text } from 'ink';

const initialView = <Text>Initializing tests...</Text>;

export const app = render(initialView);

app.unmount = jest.fn().mockImplementation(() => app.rerender(initialView));
