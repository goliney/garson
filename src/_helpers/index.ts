// Components
export { Question } from './components/question';
export { Line } from './components/line';

// Hooks
export { useKeyHandler } from './use-key-handler';
export { useEnterKeyHandler } from './use-enter-key-handler';
