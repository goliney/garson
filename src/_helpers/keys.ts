export const ARROW_UP = '\u001B[A';
export const ARROW_DOWN = '\u001B[B';
export const ARROW_LEFT = '\u001B[D';
export const ARROW_RIGHT = '\u001B[C';
export const ENTER = '\r';
export const CTRL_C = '\x03';
export const BACKSPACE = '\x08';
export const DELETE = '\x7F';
export const SPACE = ' ';
