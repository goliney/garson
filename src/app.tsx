import { Box, render } from 'ink';
import React from 'react';

export const app = render(<Box>Initializing...</Box>, {
  debug: false,
});
