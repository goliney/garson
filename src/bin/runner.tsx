import React from 'react';
import { Box, render } from 'ink';
import { GarsonConfig } from '../types';

const { rerender, unmount } = render(<Box>Initializing...</Box>, {
  debug: false,
});

export function runner({ results, prompts, action }: GarsonConfig) {
  const [{ prompt, key }, ...rest] = prompts;
  rerender(
    prompt(result => {
      rerender(<React.Fragment />); // clear the previous prompt from the screen
      const newResults = { ...results, [key]: result };
      if (rest.length === 0) {
        // it was the last prompt in a chain
        const actionResponse = action(newResults);
        if (actionResponse) {
          runner(actionResponse);
        } else {
          unmount();
        }
      } else {
        // there is another prompt
        runner({ results: newResults, prompts: rest, action });
      }
    }),
  );
}
