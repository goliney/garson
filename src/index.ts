import { KeyType, PromptCallback, Results, ActionCallback, GarsonHandlerParams } from './types';

export function garson(results?: Results) {
  return garsonHandler({ results });
}

function garsonHandler({ results = {}, prompts = [] }: GarsonHandlerParams) {
  return {
    prompt: (key: KeyType, prompt: PromptCallback) =>
      garsonHandler({ results, prompts: [...prompts, { key, prompt }] }),
    action: (action: ActionCallback) => ({
      results,
      prompts,
      action,
    }),
  };
}
