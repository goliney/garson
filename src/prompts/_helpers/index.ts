// Components
export { Question } from './components/question';

// Misc
export { useKeyHandler } from './use-key-handler';
