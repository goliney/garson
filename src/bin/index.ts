import { runner } from './runner';

const cwd = process.cwd();
let garsonConfig;

try {
  // eslint-disable-next-line import/no-dynamic-require, global-require
  garsonConfig = require(`${cwd}/garson.config.js`);
} catch (e) {
  console.error('Failed to load config', e);
}

if (garsonConfig) {
  runner(garsonConfig);
}
