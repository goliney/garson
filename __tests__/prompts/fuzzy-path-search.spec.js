import { garson, prompts } from '../../src';
import { runner } from '../../src/bin/runner';
import { app } from '../../src/app';
import { ARROW_DOWN, ARROW_UP, BACKSPACE, ENTER } from '../../src/_helpers/keys';
import { stripColorsFromLastFrame, write, waitFor } from '../helpers';

jest.mock('../../src/app');

const actionSpy = jest.fn();

const config = garson()
  .prompt(
    'file',
    prompts.fuzzyPath({
      message: 'Enter file:',
      pattern: '**',
      options: { nodir: true, cwd: '__tests__/_fixture-path/' },
    })
  )
  .action(actionSpy);

describe('Fuzzy Path Search', () => {
  beforeEach(async () => {
    actionSpy.mockClear();
    runner(config);
    await waitFor(() => {
      expect(stripColorsFromLastFrame().length).toBeGreaterThan(1);
    });
  });

  afterEach(() => {
    app.unmount();
  });

  test('Initial render', async () => {
    expect(stripColorsFromLastFrame()).toMatchSnapshot();
  });

  test('Arrow controls', async () => {
    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ foo-bar.js  .');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ elit-bar.txt  lorem');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ amet-bar.js  urna');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ ipsum-lorem.js  urna');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();
    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ ipsum-lorem.js  urna');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ amet-bar.js  urna');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ elit-bar.txt  lorem');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ foo-bar.js  .');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ bar-lorem.js  .');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ bar-lorem.js  .');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();
  });

  test('Select', async () => {
    await write(ARROW_DOWN);
    await write(ARROW_DOWN);
    await write(ENTER); // select third option
    expect(actionSpy).toHaveBeenCalledWith({
      file: {
        path: 'lorem/elit-bar.txt',
        score: null,
      },
    });
  });

  test('Search', async () => {
    await write('bar');
    expect(stripColorsFromLastFrame()).toContain('▇ bar-lorem.js  .');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    // Test how highlighting changes
    await write(BACKSPACE);
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(BACKSPACE);
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(BACKSPACE);
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write('foo');
    expect(stripColorsFromLastFrame()).toContain('▇ foo-bar.js  .');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    // Test how highlighting changes
    await write(BACKSPACE);
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(BACKSPACE);
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(BACKSPACE);
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write('lorem');
    expect(stripColorsFromLastFrame()).toContain('▇ bar-lorem.js  .');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ ipsum-lorem.js  urna');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ENTER);
    expect(actionSpy).toHaveBeenCalledWith({
      file: {
        path: 'urna/ipsum-lorem.js',
        score: {
          labelMatch: [
            {
              end: 11,
              start: 6,
            },
          ],
          score: 32768,
        },
      },
    });
  });
});
