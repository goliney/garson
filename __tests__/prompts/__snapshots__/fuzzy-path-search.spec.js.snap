// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fuzzy Path Search Arrow controls 1`] = `
"? Enter file: (Use arrow keys or start typing)
  bar-lorem.js  .
▇ foo-bar.js  .
  elit-bar.txt  lorem
  amet-bar.js  urna
  ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Arrow controls 2`] = `
"? Enter file: (Use arrow keys or start typing)
  bar-lorem.js  .
  foo-bar.js  .
▇ elit-bar.txt  lorem
  amet-bar.js  urna
  ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Arrow controls 3`] = `
"? Enter file: (Use arrow keys or start typing)
  bar-lorem.js  .
  foo-bar.js  .
  elit-bar.txt  lorem
▇ amet-bar.js  urna
  ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Arrow controls 4`] = `
"? Enter file: (Use arrow keys or start typing)
  bar-lorem.js  .
  foo-bar.js  .
  elit-bar.txt  lorem
  amet-bar.js  urna
▇ ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Arrow controls 5`] = `
"? Enter file: (Use arrow keys or start typing)
  bar-lorem.js  .
  foo-bar.js  .
  elit-bar.txt  lorem
  amet-bar.js  urna
▇ ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Arrow controls 6`] = `
"? Enter file: (Use arrow keys or start typing)
  bar-lorem.js  .
  foo-bar.js  .
  elit-bar.txt  lorem
▇ amet-bar.js  urna
  ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Arrow controls 7`] = `
"? Enter file: (Use arrow keys or start typing)
  bar-lorem.js  .
  foo-bar.js  .
▇ elit-bar.txt  lorem
  amet-bar.js  urna
  ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Arrow controls 8`] = `
"? Enter file: (Use arrow keys or start typing)
  bar-lorem.js  .
▇ foo-bar.js  .
  elit-bar.txt  lorem
  amet-bar.js  urna
  ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Arrow controls 9`] = `
"? Enter file: (Use arrow keys or start typing)
▇ bar-lorem.js  .
  foo-bar.js  .
  elit-bar.txt  lorem
  amet-bar.js  urna
  ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Arrow controls 10`] = `
"? Enter file: (Use arrow keys or start typing)
▇ bar-lorem.js  .
  foo-bar.js  .
  elit-bar.txt  lorem
  amet-bar.js  urna
  ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Initial render 1`] = `
"? Enter file: (Use arrow keys or start typing)
▇ bar-lorem.js  .
  foo-bar.js  .
  elit-bar.txt  lorem
  amet-bar.js  urna
  ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Search 1`] = `
"? Enter file: bar
▇ bar-lorem.js  .
foo-bar.js  .
amet-bar.js  urna
elit-bar.txt  lorem
ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Search 2`] = `
"? Enter file: ba
▇ bar-lorem.js  .
foo-bar.js  .
amet-bar.js  urna
elit-bar.txt  lorem
ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Search 3`] = `
"? Enter file: b
▇ bar-lorem.js  .
foo-bar.js  .
amet-bar.js  urna
elit-bar.txt  lorem
ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Search 4`] = `
"? Enter file: (Use arrow keys or start typing)
▇ bar-lorem.js  .
foo-bar.js  .
elit-bar.txt  lorem
amet-bar.js  urna
ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Search 5`] = `
"? Enter file: foo
▇ foo-bar.js  .
bar-lorem.js  .
amet-bar.js  urna
elit-bar.txt  lorem
ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Search 6`] = `
"? Enter file: fo
▇ foo-bar.js  .
bar-lorem.js  .
amet-bar.js  urna
elit-bar.txt  lorem
ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Search 7`] = `
"? Enter file: f
▇ foo-bar.js  .
bar-lorem.js  .
amet-bar.js  urna
elit-bar.txt  lorem
ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Search 8`] = `
"? Enter file: (Use arrow keys or start typing)
▇ bar-lorem.js  .
foo-bar.js  .
elit-bar.txt  lorem
amet-bar.js  urna
ipsum-lorem.js  urna"
`;

exports[`Fuzzy Path Search Search 9`] = `
"? Enter file: lorem
▇ bar-lorem.js  .
ipsum-lorem.js  urna
elit-bar.txt  lorem
foo-bar.js  .
amet-bar.js  urna"
`;

exports[`Fuzzy Path Search Search 10`] = `
"? Enter file: lorem
bar-lorem.js  .
▇ ipsum-lorem.js  urna
elit-bar.txt  lorem
foo-bar.js  .
amet-bar.js  urna"
`;
